src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs

```diff
@@ -168,5 +168,81 @@ public void MultipleTypesWithMultipleMethods()
+                //  x + (z + 2 * (r + (8 * y + 3 * (y - (5 + x))))
+                il1.Emit(OpCodes.Ldarg_1);     // push 1 MaxStack 1
+                il1.Emit(OpCodes.Ldarg_3);     // push 1 MaxStack 2
+                il1.Emit(OpCodes.Ldc_I4_2);    // push 1 MaxStack 3
+                il1.Emit(OpCodes.Ldarg_S, 4);  // push 1 MaxStack 4
+                il1.Emit(OpCodes.Ldc_I4_8);    // push 1 MaxStack 5
+                il1.Emit(OpCodes.Ldarg_2);     // push 1 MaxStack 6
+                il1.Emit(OpCodes.Mul);         // pop 2 push 1 MaxStack 5
+                il1.Emit(OpCodes.Ldc_I4_3);    // push 1 MaxStack 6
+                il1.Emit(OpCodes.Ldarg_2);     // push 1 MaxStack 7
+                il1.Emit(OpCodes.Ldc_I4_5);    // push 1 MaxStack 8
+                il1.Emit(OpCodes.Ldarg_1);     // push 1 MaxStack 9
+                il1.Emit(OpCodes.Add);         // pop 2 push 1 stack size 8 
+                il1.Emit(OpCodes.Sub);         // pop 2 push 1 stack size 7
+                il1.Emit(OpCodes.Mul);         // pop 2 push 1 stack size 6
+                il1.Emit(OpCodes.Add);         // pop 2 push 1 stack size 5
+                il1.Emit(OpCodes.Add);         // pop 2 push 1 stack size 4
+                il1.Emit(OpCodes.Mul);         // pop 2 push 1 stack size 3
+                il1.Emit(OpCodes.Add);         // pop 2 push 1 stack size 2
+                il1.Emit(OpCodes.Add);         // pop 2 push 1 stack size 1
+                il1.Emit(OpCodes.Ret);         // pop 1 stack size 0
+
+                MethodBuilder method2 = type.DefineMethod("Method2", MethodAttributes.Public, typeof(int), new Type[] { typeof(int), typeof(byte) });
+                ILGenerator il2 = method2.GetILGenerator();
+
+                // int Method2(int x, int y) =>  x + (y + 18);
+                il2.Emit(OpCodes.Ldarg_1);     // push 1 MaxStack 1
+                il2.Emit(OpCodes.Ldarg_2);     // push 1 MaxStack 2
+                il2.Emit(OpCodes.Ldc_I4_S, 8); // push 1 MaxStack 3
+                il2.Emit(OpCodes.Add);         // pop 2 push 1 stack size 2
+                il2.Emit(OpCodes.Add);         // pop 2 push 1 stack size 1
+                il2.Emit(OpCodes.Ret);         // pop 1 stack size 0
+
+                saveMethod.Invoke(ab, new object[] { file.Path });
+
+                MethodInfo getMaxStackSizeMethod = LoadILGenerator_GetMaxStackSizeMethod();
+                Assert.Equal(9, getMaxStackSizeMethod.Invoke(il1, new object[0]));
+                Assert.Equal(3, getMaxStackSizeMethod.Invoke(il2, new object[0]));
+
+                Assembly assemblyFromDisk = AssemblySaveTools.LoadAssemblyFromPath(file.Path);
+                Type typeFromDisk = assemblyFromDisk.Modules.First().GetType("MyType");
+                MethodBody body1 = typeFromDisk.GetMethod("Method1").GetMethodBody();
+                MethodBody body2 = typeFromDisk.GetMethod("Method2").GetMethodBody();
+                Assert.Equal(9, body1.MaxStackSize);
```

Comment:
Having an aggregate comparison for stack size like this is fine, but may not detect bugs at an opcode level since, for example, an off-by-one in one case may be offset by an opposite off-by-negative-one case. Should we check the current stack size after each Emit(opcode) by using reflection to investigate the current stack size in the generator (`m_curDepth`) or is this redundant in general due to other tests?

Reply:
> Should we check the current stack size after each Emit(opcode) by using reflection to investigate the current stack size in the generator (m_curDepth) or is this redundant in general due to other tests?

We can check stack size after each Emit(opcode), but might be redundant, especially in this PR only simple Emit operations covered that changes the stack linearly. My plan was keep checking the `MaxStackSize` for next PRs that add more complicated Emit overloads. It might be good to test current stack size for those Emit overloads that change the stack differently.